### ISSUE TITLE
Spring boot admin creates wrong port for health URL

### ISSUE THREAD

**Issue by Macrame** [codecentric/spring-boot-admin#1522]
I'm developing spring boot service that runs both netty port 7600 and embedded gRPC server on port 10722(using grpc-server-spring-boot-starter).
The problem is that the health URL pointing to gRPC port instead of HTTP port.

![image](https://user-images.githubusercontent.com/10723556/96559363-88690600-12ef-11eb-9f23-2fbd749a6564.png)

But another one is ok:
![image](https://user-images.githubusercontent.com/10723556/96560009-4f7d6100-12f0-11eb-83ad-701ca81bc27b.png)

I wonder how to set this behavior. My discovery is Kubernetes Service.
![image](https://user-images.githubusercontent.com/10723556/96560328-c0bd1400-12f0-11eb-95a2-3333058e18c1.png)

### Environment:
Spring Boot: 2.3.4.RELEASE
Spring Cloud: Hoxton.SR8
Kubernetes: 1.19.3
Istio 1.7.3

Spring Boot Admin 2.3.0

**Comment by Macrame** [comment]
I changed the name of port in Kubernetes svc to gRPC but the health port is still the wrong one. How to determine which port is the one for health?
```
Name:              mars-user-service-v1-0
Namespace:         mars
Labels:            app=mars-user-service-v1-0
                   app.kubernetes.io/managed-by=Helm
                   gRPC.port=10682
                   user.name=spring-boot-admin
                   user.password=galaxy_system
                   version=v1-0
Annotations:       meta.helm.sh/release-name: mars-user-service-v1-0
                   meta.helm.sh/release-namespace: mars
Selector:          app.kubernetes.io/managed-by=Helm,app=mars-user-service-v1-0
Type:              ClusterIP
IP:                10.0.6.147
Port:              http  9200/TCP
TargetPort:        9200/TCP
Endpoints:         10.1.121.36:9200
Port:              grpc  10682/TCP
TargetPort:        10682/TCP
Endpoints:         10.1.121.36:10682
Session Affinity:  None
Events:            <none>


Name:              mars-control-strategy-service-v1-0
Namespace:         mars
Labels:            app=mars-control-strategy-service-v1-0
                   app.kubernetes.io/managed-by=Helm
                   gRPC.port=10722
                   user.name=spring-boot-admin
                   user.password=galaxy_system
                   version=v1-0
Annotations:       meta.helm.sh/release-name: mars-control-strategy-service-v1-0
                   meta.helm.sh/release-namespace: mars
Selector:          app.kubernetes.io/managed-by=Helm,app=mars-control-strategy-service-v1-0
Type:              ClusterIP
IP:                10.0.6.50
Port:              http  7600/TCP
TargetPort:        7600/TCP
Endpoints:         10.1.121.40:7600
Port:              grpc  10722/TCP
TargetPort:        10722/TCP
Endpoints:         10.1.121.40:10722
Session Affinity:  None
Events:            <none>
```

**Comment by joshiste** [comment]
When using the k8s discovery SBA has the following strategy to determine the port:
1. use `port.management` from metadata if present.
2. use the port supplied by the discovery client (spring-cloud-kubernetes)

In order to make it work for you either:

a) map the actuator endpoints to a kubernetes  port called `management`
or
b) Add a custom bean subclassing ` KubernetesServiceInstanceConverter` in your SBA server taking other metadata into account

**Comment by Macrame** [comment]
Thanks a lot. I'll try it.

**Comment by Macrame** [comment]
@joshiste  But the two applications have different behavior: One uses the http port while the other uses the gRPC port. It looks like a random port strategy.

**Comment by joshiste** [comment]
> If the spring-cloud-kubernetes supplies two ports, will SBA pick up the first one or random one regardless of the name of port? Sorry for so detail question.

SBA will use the port whichever is picked by spring-cloud-kubernetes ...

**Comment by Macrame** [comment]
A better way is  specifying the name of the primary port.
```
spring:
    cloud:
        kubernetes:
            enabled: true
            loadbalancer:
                enabled: true
            discovery:
                primary-port-name: <The target port>
```
[Refer to ](https://github.com/spring-cloud/spring-cloud-kubernetes/issues/341)

**Comment by martin-schaefer** [comment]
Hi Macrame,
It seems I have the same issue. But I noticed that the problem only arises when the Istio sidecars are injected into the Spring Boot Pod.
I have exposed the Actuators on port 8010 and named this port "http-management" in the Kubernetes service object. It has also the port 80 exposed as "http-service". I configured Spring Cloud Kubernetes in Spring Boot Admin accordingly.
```
spring:
  cloud:
    kubernetes:
      discovery:
        primary-port-name: http-management

```
If I run this without Istio injection, Spring Boot Admin takes port 8010 and it works fine. But as soon as I activate Istio injection Spring Boot Admin takes port 80 and it doesnt work anymore.
My guess is, that Istio Injection has some side effects on Spring Cloud Kubernetes discovery. But I could not find a way around this except to disable Istio Injection.

Does your problem disappear when disabling Istio?

**Comment by Macrame** [comment]
@martin-schaefer Could you paste the output of this command `kubectl describe ep <Your Service_Name>`? Please note the `Ports` under `Subsets` and confirm there are two ports and the name should be correct.

**Comment by Macrame** [comment]
Another possibility, have you used @EnableScheduling on the main application class in Spring Boot Admin? Otherwise, you'd better recreate the Pod and Service since when the admin server starts up it finds the list of running applications and monitors their health status but it does not know of any new instances that are started. See #643

**Comment by martin-schaefer** [comment]
Hi @Macrame 
After further analysis, I found the root cause of the problem: I am using the Spring Cloud Kubernetes feature to read ConfigMaps as PropertySource. But for some reason this breaks when istio sidecar injection is activated. So Spring Boot Admin did not get the primary-port-name form the configuration and just took the first one (port 80).
When I set the primary-port-name direct at vm startup with parameter -Dspring.cloud.kubernetes.discovery.primary-port-name=http-management it works fine.

So the problem is related to the combination of Spring Cloud Kubernetes and istio sidecar injection. I'll probably open a ticket at Spring Cloud Kubernetes project.
